Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm

```diff
@@ -5463,43 +5451,30 @@ - (void)beginSelectionInDirection:(UITextDirection)direction completionHandler:(
 }
 
 - (void)updateSelectionWithExtentPoint:(CGPoint)point completionHandler:(void (^)(BOOL endIsMoving))completionHandler
+{
+    [self updateSelectionWithExtentPoint:point withBoundary:UITextGranularityCharacter completionHandler:completionHandler];
+}
+
+- (void)updateSelectionWithExtentPoint:(CGPoint)point withBoundary:(UITextGranularity)granularity completionHandler:(void (^)(BOOL selectionEndIsMoving))completionHandler
 {
     logTextInteraction(__PRETTY_FUNCTION__, self.textInteractionLoupeGestureRecognizer, point);
 
     _autocorrectionContextNeedsUpdate = YES;
 
-    auto hasRecognizedOrEnded = [](UIGestureRecognizer *gestureRecognizer) {
-        switch (gestureRecognizer.state) {
-        case UIGestureRecognizerStateBegan:
-        case UIGestureRecognizerStateChanged:
-        case UIGestureRecognizerStateEnded:
-            return true;
-        case UIGestureRecognizerStatePossible:
-        case UIGestureRecognizerStateCancelled:
-        case UIGestureRecognizerStateFailed:
-            return false;
-        }
-        ASSERT_NOT_REACHED();
-        return false;
-    };
-
-    auto triggeredByFloatingCursor = !hasRecognizedOrEnded(self.textInteractionLoupeGestureRecognizer)
+    if (granularity == UITextGranularityCharacter && !_usingMouseDragForSelection) {
+        auto triggeredByFloatingCursor = !self.textInteractionLoupeGestureRecognizer._wk_hasRecognizedOrEnded
 #if HAVE(UIKIT_WITH_MOUSE_SUPPORT)
-        && !hasRecognizedOrEnded([_mouseInteraction mouseTouchGestureRecognizer])
+            && ![_mouseInteraction mouseTouchGestureRecognizer]._wk_hasRecognizedOrEnded
 #endif
-        && !hasRecognizedOrEnded(self.textInteractionTapGestureRecognizer);
+            && !self.textInteractionTapGestureRecognizer._wk_hasRecognizedOrEnded;
 
-    auto respectSelectionAnchor = triggeredByFloatingCursor ? WebKit::RespectSelectionAnchor::Yes : WebKit::RespectSelectionAnchor::No;
-    _page->updateSelectionWithExtentPoint(WebCore::IntPoint(point), self._hasFocusedElement, respectSelectionAnchor, [selectionHandler = makeBlockPtr(completionHandler)](bool endIsMoving) {
-        selectionHandler(endIsMoving);
-    });
-}
-
-- (void)updateSelectionWithExtentPoint:(CGPoint)point withBoundary:(UITextGranularity)granularity completionHandler:(void (^)(BOOL selectionEndIsMoving))completionHandler
-{
-    logTextInteraction(__PRETTY_FUNCTION__, self.textInteractionLoupeGestureRecognizer, point);
+        auto respectSelectionAnchor = triggeredByFloatingCursor ? WebKit::RespectSelectionAnchor::Yes : WebKit::RespectSelectionAnchor::No;
+        _page->updateSelectionWithExtentPoint(WebCore::IntPoint(point), self._hasFocusedElement, respectSelectionAnchor, [selectionHandler = makeBlockPtr(completionHandler)](bool endIsMoving) {
```

Comment:
Do we need to be careful of `bool` / `BOOL` here?

(completionHandler takes a `BOOL`, but `endIsMoving` is a `bool`)

Reply:
Good catch — though, this implicit conversion was also in existing code... I think the Catalyst build issue fixed in [272193@main](https://commits.webkit.org/272193@main) only applies to implicit conversion when determining the return value of a block.

I'll add a `static_cast<BOOL>` here to make the conversion explicit.

Thanks for the review!